polynote-server/src/main/scala/polynote/server/NotebookManager.scala

```diff
@@ -0,0 +1,65 @@
+package polynote.server
+
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+import cats.Monad
+import cats.effect.{ContextShift, Fiber, IO}
+import cats.effect.concurrent.Semaphore
+import polynote.server.repository.NotebookRepository
+
+/**
+  * This is how the sessions interact with notebooks. The sessions don't think about writing changes etc. They'll
+  * open a [[SharedNotebook]] and from there get a [[NotebookRef]] in exchange for an output queue. Their interactions
+  * with the notebook will go through the NotebookRef.
+  */
+abstract class NotebookManager[F[_]](implicit F: Monad[F]) {
```

Comment:
All the methods defined here are also defined in the `NotebookRepository` trait but there's no relationship between them, is that intentional?

(incidentally, it looks like we have `(implicit F: Monad[F])` here but not in `NotebookRepository`. I guess what that does is allow us, e.g., to create a `NotebookManager[Notebook]` assuming we have an implicit `Notebook => IO[Notebook]` available?)

Reply:
The relationship is that `NotebookManager` is expected to have a `NotebookRepository` to talk to – nobody else gets to talk to the repository.

The reason for the context bound is just so I can use `map` and `flatMap` in some default implementations. Those could just as well be abstract and left to the concrete class, and it wouldn't need a context bound.

Reply:
P.S. Maybe `NotebookRepository` really wants to be `NotebookFormat` or something... not totally sure yet. But `NBManager` has to have those APIs because it's all that `SocketSession` gets to talk to.